marlowe-spec-test/src/Marlowe/Spec/Core/Serialization/Json.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Marlowe.Spec.Core.Serialization.Json where
+
+import Control.Applicative ((<|>))
+import Data.Aeson.Types (Result(..), ToJSON(..), FromJSON(..))
+import Data.Aeson (object, (.=), (.:), withObject)
+import qualified Data.Aeson.Types as JSON
+import Data.Text as T
+import Data.Proxy (Proxy(..))
+import MarloweCoreJson
+import GHC.Stack (HasCallStack)
+import Test.Tasty (TestTree, testGroup)
+import Marlowe.Spec.Interpret (Response (..), InterpretJsonRequest, exactMatch, Request (..), testResponse)
+import Marlowe.Spec.TypeId (TypeId(..), HasTypeId (..))
+import Test.Tasty.Providers (TestName)
+import Test.Tasty.HUnit (Assertion, assertBool)
+import qualified SemanticsTypes as C
+
+
+data SerializationResponse transport
+  = SerializationSuccess transport
+  | UnknownType TypeId
+  | SerializationError String
+  deriving (Eq)
+
+instance ToJSON (SerializationResponse JSON.Value) where
+  toJSON (SerializationSuccess result) = object
+    [ "serialization-success" .= result
+    ]
+  toJSON (UnknownType t) = object
+    [ "unknown-type" .= toJSON t
+    ]
+  toJSON (SerializationError err) = object
+    [ "serialization-error" .= JSON.String (T.pack err)
+    ]
+
+instance FromJSON (SerializationResponse JSON.Value) where
+  parseJSON = withObject "SerializationResponse" $
+      \v -> asSuccess v <|> asUnknownType v <|> asError v
+    where
+    asSuccess v = SerializationSuccess <$> v .: "serialization-success"
+    asUnknownType v = UnknownType <$> v .: "unknown-type"
+    asError v = SerializationError <$> v .: "serialization-error"
+
+localJsonRoundtripSerialization :: TypeId -> JSON.Value -> SerializationResponse JSON.Value
+localJsonRoundtripSerialization (TypeId _ p) v = roundtrip p
+    where
+    roundtrip :: forall a. ToJSON a => FromJSON a => Proxy a -> SerializationResponse JSON.Value
+    roundtrip _  = case JSON.fromJSON v :: Result a of
+            Error str -> SerializationError str
+            Success c ->  SerializationSuccess $ JSON.toJSON c
+
+
+tests :: InterpretJsonRequest -> TestTree
+tests i = testGroup "Json Serialization"
+  [ roundtripExampleTest i "Bound example" condExample
+  , valueTests i
+  , observationTests i
+  , invalidType i
+  ]
+
+valueTests :: InterpretJsonRequest -> TestTree
+valueTests i = testGroup "Value examples"
+  [ roundtripExampleTest i "Constant" constantExample
+  , roundtripExampleTest i "Interval start" intervalStartExample
+  , roundtripExampleTest i "Interval end" intervalEndExample
+  , roundtripExampleTest i "Add" addExample
+  , roundtripExampleTest i "Sub" subExample
+  , roundtripExampleTest i "Mul" mulExample
+  , roundtripExampleTest i "Div" divExample
+  , roundtripExampleTest i "Negate" negateExample
+  -- , roundtripExampleTest i "Choice value" choiceValueExample
```

Comment:
These are commented until I make `SCP-4697: Implement GenerateRandomValue for Token and Address`